### PR TITLE
[DEV-6053] Simplify Award Delta View by Moving Logic to MVs

### PR DESCRIPTION
**Pull request by collinwr**
**Description:**
This PR moves joins and calculations from the `award_delta_view.sql` file to the award search materialized views.

**Technical details:**
The `award_delta_view` still has the same columns and the data should match the previous version.

Three main areas updated:
- Award Delta View json files
- `base_award_search.py`
- `award_delta_view.py`

It moves joins and aggregations down, so the delta view simply selects fields straight from `vw_es_award_search`. 

To ensure no changes to the data were made, a sql script compared contents of the original `award_delta_view` with the updated version (pulling from new matviews) locally.

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. N/A - API documentation updated
3. [ ] Necessary PR reviewers:
    - [x] Backend
    - [ ] Frontend <OPTIONAL>
    - [ ] Operations <OPTIONAL>
    - [ ] Domain Expert <OPTIONAL>
4. [x] Matview impact assessment completed
5. N/A - Frontend impact assessment completed
6. [x] Data validation completed
7. [x] Appropriate Operations ticket(s) created - [DEV-6421](https://federal-spending-transparency.atlassian.net/browse/DEV-6421)
8. [x] Jira Ticket [DEV-6053](https://federal-spending-transparency.atlassian.net/browse/DEV-6053):
    - [x] Link to this Pull-Request
    - [ ] Performance evaluation of affected (API | Script | Download)
    - [x] Before / After data comparison

**Area for explaining above N/A when needed:**
```
No changes to API. No changes to front end
```

### CODE REVIEW

File: usaspending_api/database_scripts/etl/award_delta_view.sql [hunk 1]
```diff
@@ -183,114 +206,16 @@ SELECT
...
-  INNER JOIN financial_accounts_by_awards faba ON (taa.treasury_account_identifier = faba.treasury_account_id)
-  INNER JOIN federal_account fa ON (taa.federal_account_id = fa.id)
-  INNER JOIN toptier_agency agency ON (fa.parent_toptier_agency_id = agency.toptier_agency_id)
-  WHERE
-    faba.award_id IS NOT NULL
-  GROUP BY
-    faba.award_id
-) TREASURY_ACCT ON (TREASURY_ACCT.award_id = vw_es_award_search.award_id)
-;
+;
```
Comment (tony-sappe) [inline usaspending_api/database_scripts/etl/award_delta_view.sql:221]: missing trailing newline

File: usaspending_api/database_scripts/etl/award_delta_view.sql [hunk 2]
```diff
@@ -20,26 +20,27 @@ SELECT
...
+          '","hash":"","levels":""}'	
+        )	
+    ELSE	
+      CONCAT(	
+        '{"name":"', vw_es_award_search.recipient_name,	
+        '","unique_id":"', vw_es_award_search.recipient_unique_id,	
+        '","hash":"', vw_es_award_search.recipient_hash,	
+        '","levels":"', vw_es_award_search.recipient_levels, '"}'	
+      )	
   END AS recipient_agg_key,
```
Comment (tony-sappe) [inline usaspending_api/database_scripts/etl/award_delta_view.sql:44]: I thought the goal was the move even these fields into the matviews

**Review by tony-sappe — CHANGES_REQUESTED** [review]

File: usaspending_api/search/models/base_award_search.py [hunk 3]
```diff
@@ -89,5 +105,11 @@ class BaseAwardSearchModel(models.Model):
     naics_code = models.TextField()
     naics_description = models.TextField()
 
+    tas_paths = models.TextField()
+    tas_components = models.TextField()
+    disaster_emergency_fund_codes = models.TextField()
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/search/models/base_award_search.py:110]: Just double checked the database. I see that Array fields are notated by a `_<type>`. I fixed this

File: usaspending_api/search/models/base_award_search.py [hunk 4]
```diff
@@ -62,22 +65,35 @@ class BaseAwardSearchModel(models.Model):
...
     pop_county_name = models.TextField()
+    pop_country_population = models.IntegerField()
     pop_city_code = models.TextField()
     pop_zip5 = models.TextField()
+    pop_congressional_population = models.IntegerField()
     pop_congressional_code = models.TextField()
     pop_city_name = models.TextField()
 
+    cfda_program_title = models.TextField()
+    cfda_id = models.TextField()
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/search/models/base_award_search.py:95]: Checked the database. IntegerField is correct. I have fixed it

File: usaspending_api/search/models/base_award_search.py [hunk 5]
```diff
@@ -31,6 +33,7 @@ class BaseAwardSearchModel(models.Model):
 
     recipient_hash = models.UUIDField()
     recipient_name = models.TextField()
+    recipient_levels = models.TextField()
```
[the remark this answers is not in the document]
Comment (collinwr) [reply usaspending_api/search/models/base_award_search.py:36]: ArrayField is correct. Fixed this

File: usaspending_api/database_scripts/etl/award_delta_view.sql (hunk 1 above)
Comment (collinwr) [reply usaspending_api/database_scripts/etl/award_delta_view.sql:221]: Added

File: usaspending_api/database_scripts/etl/award_delta_view.sql [hunk 6]
```diff
@@ -20,27 +16,13 @@ SELECT
...
-    ELSE
-      CONCAT(
-        '{"name":"', vw_es_award_search.recipient_name,
-        '","unique_id":"', vw_es_award_search.recipient_unique_id,
-        '","hash":"', recipient_profile.recipient_hash,
-        '","levels":"', recipient_profile.recipient_levels, '"}'
-      )
-  END AS recipient_agg_key,
+  vw_es_award_search.recipient_hash,
+  vw_es_award_search.recipient_levels,
```
Comment (tony-sappe) [inline usaspending_api/database_scripts/etl/award_delta_view.sql:24]: is `recipient_levels` needed in the view?

File: usaspending_api/database_scripts/etl/award_delta_view.sql [hunk 7]
```diff
@@ -66,47 +48,41 @@ SELECT
...
-      )
-    ELSE NULL
-  END AS funding_subtier_agency_agg_key,
+  vw_es_award_search.funding_toptier_agency_agg_key,
+  vw_es_award_search.funding_subtier_agency_agg_key,
 
   vw_es_award_search.recipient_location_country_code,
   vw_es_award_search.recipient_location_country_name,
   vw_es_award_search.recipient_location_state_code,
+  vw_es_award_search.recipient_location_state_name,
```
Comment (tony-sappe) [inline usaspending_api/database_scripts/etl/award_delta_view.sql:57]: Appears that a number of new fields are being added. Are they necessary for Elasticsearch?

File: usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json [hunk 8]
```diff
@@ -9,6 +9,12 @@
     "  awards.category,",
     "  awards.type,",
     "  awards.type_description,",
+    "  awards.generated_unique_award_id,",
+    "  CASE",
+    "    WHEN awards.piid IS NOT NULL THEN awards.piid  -- contracts. Did it this way to easily handle IDV contracts",
```
Comment (tony-sappe) [inline usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json:14]: `piid` won't be null for contracts or IDVs The case statement can be removed

File: usaspending_api/database_scripts/matview_generator/mv_directpayment_award_search.json [hunk 9]
```diff
@@ -9,6 +9,12 @@
     "  awards.category,",
     "  awards.type,",
     "  awards.type_description,",
+    "  awards.generated_unique_award_id,",
+    "  CASE",
+    "    WHEN awards.fain IS NOT NULL THEN awards.fain",
+    "    ELSE awards.uri",
+    "  END AS display_award_id,",
```
Comment (tony-sappe) [inline usaspending_api/database_scripts/matview_generator/mv_directpayment_award_search.json:16]: Can change display_award_id to `COALESCE(awards.fain, awards.uri)`

File: usaspending_api/search/models/base_award_search.py [hunk 10]
```diff
@@ -19,6 +19,8 @@ class BaseAwardSearchModel(models.Model):
     category = models.TextField()
```
Comment (tony-sappe) [inline usaspending_api/search/models/base_award_search.py:19]: missing *_agg_key fields in the base model

**Review by tony-sappe — CHANGES_REQUESTED** [review]
Few minor comments. I haven't compared the original SQL to the new SQL yet

File: usaspending_api/search/models/base_award_search.py [hunk 11]
```diff
@@ -19,6 +19,9 @@ class BaseAwardSearchModel(models.Model):
...
     type = models.TextField()
     type_description = models.TextField()
```
Comment (tony-sappe) [inline usaspending_api/search/models/base_award_search.py:21]: Discrepancies between materialized view definition and the base class. (looking at `mv_directpayment_award_search`) 

`pop_country_population` vs `pop_county_population` (which one? or both?)
missing `total_covid_obligation`

image

Comment (tony-sappe) [reply usaspending_api/search/models/base_award_search.py:21]: Verified the materialized views schema match each other

File: usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json [hunk 12]
```diff
@@ -50,27 +69,57 @@
...
+    "      )",
+    "    ELSE NULL",
+    "  END AS funding_subtier_agency_agg_key,",
     "",
     "  rl_country_lookup.country_code AS recipient_location_country_code,",
     "  rl_country_lookup.country_name AS recipient_location_country_name,",
     "  transaction_fpds.legal_entity_state_code AS recipient_location_state_code,",
+    "  RL_STATE_LOOKUP.name AS recipient_location_state_name,",
+    "  RL_STATE_LOOKUP.fips AS recipient_location_state_fips,",
+    "  RL_STATE_POPULATION.latest_population AS recipient_location_state_population,",
```
Comment (tony-sappe) [inline usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json:96]: I don't see these fields in the original view. What will they help?

File: usaspending_api/database_scripts/etl/award_delta_view.sql (hunk 6 above)
Comment (collinwr) [reply usaspending_api/database_scripts/etl/award_delta_view.sql:24]: No, it looks like it was only previously included to build the agg keys. I will remove it from the mvs, django model, and delta view

File: usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json (hunk 8 above)
Comment (collinwr) [reply usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json:14]: `display_award_id` is now always `awards.piid` in the `mv_idv_award_search`

Comment (collinwr) [reply usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json:14]: `display_award_id` is now always `awards.piid` in `mv_idv_award_search`

File: usaspending_api/database_scripts/matview_generator/mv_directpayment_award_search.json (hunk 9 above)
Comment (collinwr) [reply usaspending_api/database_scripts/matview_generator/mv_directpayment_award_search.json:16]: Updated in other, directpayments, grant, and loan

File: usaspending_api/search/models/base_award_search.py (hunk 10 above)
Comment (collinwr) [reply usaspending_api/search/models/base_award_search.py:19]: These have been added

File: usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json (hunk 8 above)
Comment (tony-sappe) [reply usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json:14]: It looks the same:

```sql
    CASE
    WHEN awards.piid IS NOT NULL THEN awards.piid  -- contracts. Did it this way to easily handle IDV contracts
    ELSE NULL
    END AS display_award_id,
```

It should be reduced to `awards.piid AS display_award_id` for both of the procurement materialized views

File: usaspending_api/database_scripts/etl/award_delta_view.sql (hunk 7 above)
Comment (collinwr) [reply usaspending_api/database_scripts/etl/award_delta_view.sql:57]: Removed `recipient_location_state_name`, `recipient_location_state_fips`, `recipient_location_state_population`, `recipient_location_county_population`, `recipient_location_congressional_population`, `pop_state_name`, `pop_state_fips`, `pop_state_population`, `pop_county_population`, and `pop_congressional_population`

File: usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json (hunk 12 above)
Comment (collinwr) [reply usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json:96]: Removed `recipient_location_state_name`, `recipient_location_state_fips`, `recipient_location_state_population`, `recipient_location_county_population`, `recipient_location_congressional_population`, `pop_state_name`, `pop_state_fips`, `pop_state_population`, `pop_county_population`, and `pop_congressional_population`. (Same as above in the delta view)

File: usaspending_api/search/models/base_award_search.py (hunk 11 above)
Comment (collinwr) [reply usaspending_api/search/models/base_award_search.py:21]: Removed `pop_country_population` from model and `pop_county_population` from MVs

File: usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json (hunk 8 above)
Comment (collinwr) [reply usaspending_api/database_scripts/matview_generator/mv_contract_award_search.json:14]: Made this change for `mv_contract_award_search` as well

**Review by tony-sappe — APPROVED** [review]
Approving code changes. There might be a last-minute tweak if manual programatic comparison tests unearth anything. PR description could use a little TLC before merging

**Pull request merged**